src/Apropos/Gen.hs

```diff
@@ -38,32 +41,64 @@ import Data.String (fromString)
 import Hedgehog (PropertyT)
 import Hedgehog qualified as H
 import Hedgehog.Gen qualified as HGen
-import Hedgehog.Internal.Gen qualified as HIG
-import Hedgehog.Internal.Tree qualified as HIT
 import Hedgehog.Range qualified as HRange
 
-forAll :: Show a => Gen a -> PropertyT IO a
-forAll g = do
-  (ee, labels) <- H.forAll $ runWriterT (runExceptT $ gen g)
-  mapM_ (H.label . fromString) labels
+forAllWithRetries :: forall a. Show a => Int -> Gen a -> PropertyT IO (Either GenException a)
+forAllWithRetries retries g = go 0
+  where
+    go :: Int -> PropertyT IO (Either GenException a)
+    go l = do
+      res <- forAll $ scale (2 * l +) g
+      case res of
+        Left Retry ->
+          if l > retries
+            then pure $ Left Retry
+            else go (l + 1)
+        Left err -> pure $ Left err
+        Right so -> pure $ Right so
+
+errorHandler :: Either GenException a -> PropertyT IO a
+errorHandler ee =
   case ee of
     Left Retry -> H.footnote "retry limit reached" >> H.discard
     Left (GenException err) -> H.footnote err >> H.failure
     Right a -> pure a
 
-forAllWith :: forall a. (a -> String) -> Gen a -> PropertyT IO a
-forAllWith s g = do
-  (ee, labels) <- H.forAllWith sh $ runWriterT (runExceptT $ gen g)
+forAll :: Show a => Gen a -> PropertyT IO (Either GenException a)
+forAll = interleaved . gen2Interleaved
+
+forAllWith :: forall a. (a -> String) -> Gen a -> PropertyT IO (Either GenException a)
+forAllWith s = interleavedWith s . gen2Interleaved
+
+forAllWithG :: forall a. (a -> String) -> Generator a -> PropertyT IO (Either GenException a)
+forAllWithG s g = do
+  (ee, labels) <- H.forAllWith sh $ runWriterT (runExceptT g)
   mapM_ (H.label . fromString) labels
-  case ee of
-    Left Retry -> H.footnote "retry limit reached" >> H.discard
-    Left (GenException err) -> H.footnote err >> H.failure
-    Right a -> pure a
+  pure ee
   where
     sh :: (Either GenException a, Set String) -> String
     sh (Left e, _) = show e
     sh (Right a, _) = s a
 
+interleaved :: forall a. Show a => Interleaved a -> PropertyT IO (Either GenException a)
+interleaved = interleavedWith show
+
+interleavedWith :: forall a. (a -> String) -> Interleaved a -> PropertyT IO (Either GenException a)
+interleavedWith s m = do
+  res <- forAllWithG sh $ gSteps m
+  case res of
+    Right (Right a) -> pure $ Right a
+    Right (Left c) -> do
+      pres <- pSteps c
+      case pres of
+        Right a -> pure $ Right a
+        Left pc -> interleavedWith s pc
+    Left err -> pure $ Left err
+  where
+    sh :: Either (Interleaved a) a -> String
+    sh (Left _) = "This is unexpected. Generator interleaving failed. Please contact a customer support representative."
+    sh (Right a) = s a
```

Comment:
This does the interleaving of `Gen` computations with "lifted" embedded `PropertyT IO` computations. We bind as many `Gen` computations as we can together then run them with `forAllWithG`. Then if we do the same with `PropertyT IO` - bind as many as we can together and run them. This works recursively to interleave the two contexts and allows us to embed `PropertyT IO` in `Gen`.